### PR TITLE
re-enable the module resolution build tests

### LINKED ISSUE #71: Add build tests for module-resolution playground
These tests pass in dev but not when the build output is tested. Needs investigating.

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
resolves #71

### CODE REVIEW

File: playground/module-resolution/__tests__/module-resolution.spec.ts
Comment (dario-piotrowicz) [inline playground/module-resolution/__tests__/module-resolution.spec.ts:137]: the test here is testing the vite error overlay which is indeed a dev only vite feature so this actually should be skipped for builds

**Comment by petebacondarwin** [comment]
What are these assets modules that are being included?

**Comment by dario-piotrowicz** [comment]
> What are these assets modules that are being included?

the transpiled js that `vite build` generates

For example in the module resolution app this is what we get:
Screenshot 2024-12-02 at 11 41 38

And this is for example the `aliasing.ts` couple of files:
Screenshot 2024-12-02 at 11 42 42

**Comment by petebacondarwin** [comment]
Oh! They are the result of the `client` build, right?

**Comment by dario-piotrowicz** [comment]
> Oh! They are the result of the `client` build, right?

No, they are part of the worker build as you can see they are under `dist/worker`, while client files would be under `dist/client`

They are necessary for the worker to run since its imports get rewritten to use these files, for example this is what we get when we build the module resolution worker:
Screenshot 2024-12-02 at 12 51 57

This being the source file:
Screenshot 2024-12-02 at 12 52 15

**Comment by petebacondarwin** [comment]
Oh OK. I was very confused by the reference to "assets". Although Vite might call them assets, I think it is confusing to refer to them as that in this PR. They are really just additional modules needed by the server-side Worker. Right?

**Comment by dario-piotrowicz** [comment]
> Oh OK. I was very confused by the reference to "assets". Although Vite might call them assets, I think it is confusing to refer to them as that in this PR. They are really just additional modules needed by the server-side Worker. Right?

yeah I totally agree "assets" is pretty confusing 😓 

By the way, this is also part of the vite standard naming, my understanding is that they call these "generated assets": https://vite.dev/config/build-options.html#build-assetsdir

Given that Vite calls them assets it felt safer/more natural for me to also call them such so that we don't diverge from Vite's naming. I'm afraid that diverging from Vite's naming can in the long run make such code more confusing/difficult to keep track of, but if you definitely prefer us using a different name here I'd be fine with it 🙂 (and in that case, do you have any suggestion?)

**Comment by petebacondarwin** [comment]
I think it was more a case of the PR title and description. Fine to call them assets but can we distinguish them by environment? Perhaps ssr assets, rather than client assets?

**Comment by dario-piotrowicz** [comment]
> I think it was more a case of the PR title and description. Fine to call them assets but can we distinguish them by environment? Perhaps ssr assets, rather than client assets?

does workers assets sound good? 🙂

File: packages/vite-plugin-cloudflare/src/miniflare-options.ts
Comment (petebacondarwin) [reply packages/vite-plugin-cloudflare/src/miniflare-options.ts:479]: My main concern is how that fits with "Workers Assets" projects...

Comment (dario-piotrowicz) [reply packages/vite-plugin-cloudflare/src/miniflare-options.ts:479]: oh yeah 😓

**Comment by dario-piotrowicz** [comment]
@petebacondarwin how about now? https://github.com/flarelabs-net/vite-plugin-cloudflare/pull/80/commits/daa5e4dbdb0a3f7cf4eca710dd68f115b0f3e4f9 🙂

**Review by petebacondarwin — COMMENTED** [review]
I think I am good with this. What about non-JS modules (e.g. WASM)?
They will need to be provided to Miniflare too, right?

Would another approach be to implement the module fallback service?

**Comment by dario-piotrowicz** [comment]
> I think I am good with this. What about non-JS modules (e.g. WASM)? They will need to be provided to Miniflare too, right?

yeah I would assume so, we can investigate it as a followup if that's ok?

> Would another approach be to implement the module fallback service?

ah yeah! nice idea! 😃 shall I change my current implementation to use that instead?

**Comment by petebacondarwin** [comment]
> ah yeah! nice idea! 😃 shall I change my current implementation to use that instead? 

Up to you. If this is blocking and we need something now, we can land this and follow up - and I can approve this.
If you want to give the fallback approach a try first, please go ahead.

**Comment by dario-piotrowicz** [comment]
I've converted the PR to draft because:
 - we agreed that this PR can wait for @jamesopstad repo restructuring
 - I'll try to use the module fallback service as suggested by @petebacondarwin

**Comment by dario-piotrowicz** [comment]
@petebacondarwin I've looked into using the module fallback here, but I don't think that it is worth it 😕

The fact is that, as we already know, there is only a single module fallback (instead of one per-worker), so since we're supporting multiworkers, in its implementation we'd have to discern what worker is triggering the import and based on that read the imported file from the appropriate generated assets directory

That would be possible/relatively simple since in the module fallback we do get a `referrer` value that indicates what files in triggering the import (e.g. `/dist/worker/index.js`), however this extra of logic makes me think that the current solution of just reading the files per-worker beforehand and adding the extra modules to the miniflare options is cleaner and simpler (and potentially more robust), so unless we would get some clear benefits from using the module fallback service I am thinking that we should stick with the current approach 🤔

**Comment by jamesopstad** [comment]
I'm just adding comments for my PR now so it will be ready soon but I'm just jumping in to say that this should hopefully just work once it's landed. The `wrangler.json` includes module rules in the output and these are read by Miniflare.

**Review by jamesopstad — COMMENTED** [review]
If we remove the following code from playground/module-resolution/src/index.ts then it's possible to remove the changes to `getEntryModule` and the new `getWorkerAdditionalJsModules` function and all the build tests will still pass:

```ts
if (path === '/@non-existing/pkg') {
  const { test } = await import('@non-existing/pkg');
  return test();
}
```

This is because Miniflare will use the `modulesRules` to find the additional modules. It will otherwise fail on '@non-existing/pkg` because it doesn't exist.

I would rather stick to this simpler approach, firstly because it doesn't add extra complexity but more importantly because it's closer to production and validates that the `rules` are specified correctly. What do you think?

**Comment by dario-piotrowicz** [comment]
> If we remove the following code from `playground/module-resolution/src/index.ts` then it's possible to remove the changes to `getEntryModule` and the new `getWorkerAdditionalJsModules` function and all the build tests will still pass:
> 
> ```ts
> if (path === '/@non-existing/pkg') {
>   const { test } = await import('@non-existing/pkg');
>   return test();
> }
> ```
> 
> This is because Miniflare will use the `modulesRules` to find the additional modules. It will otherwise fail on `@non-existing/pkg` because it doesn't exist.
> 
> I would rather stick to this simpler approach, firstly because it doesn't add extra complexity but more importantly because it's closer to production and validates that the `rules` are specified correctly. What do you think?

good callout, I didn't know miniflare would fine additional modules on its own

if that's the case I agree that we should leave it to miniflare 👍 

but what happens if you try to import the non-existing package? we do need to support that and show an acceptable error to the user 🤔

**Comment by jamesopstad** [comment]
> but what happens if you try to import the non-existing package? we do need to support that and show an acceptable error to the user 🤔

Yes, I think we need to understand more about Miniflare's behaviour here. The fact that it errors before a request is received seems to indicate that it scans the file for async imports somehow.

**Comment by dario-piotrowicz** [comment]
converted to draft as I want to understand the non-existing package issue (and hopefully fix it) before merging the PR

(unless we think it'd be beneficial to merge this sooner?)

**Comment by dario-piotrowicz** [comment]
> converted to draft as I want to understand the non-existing package issue (and hopefully fix it) before merging the PR
> 
> (unless we think it'd be beneficial to merge this sooner?)

As discussed let's just remove the non-existing test entirely

<details>

<summary>for future reference</summary>

The issue is that, in preview, if Miniflare fails to detect the package it errors before even starting completely breaking the application:
Screenshot 2024-12-06 at 15 05 33

Note that this error message is valid, so the behavior here is indeed correct (the only issue is that it prevents us from testing anything else in build/preview for the app)

Trying to hide the literal string from Miniflare doesn't help either:
Screenshot 2024-12-06 at 15 09 36

So, since this was only needed to test that locally in dev mode a nice vite error overlay presents the error, we decided that it is ok to just remove this for now

(and If we do want to test the overlay error in the future we can always add a new dev-only playground app just for this sort of things)

</details>

**Review by jamesopstad — APPROVED** [review]
Looks good!

**Pull request merged**
